** 1. XOR Function**

In class we attempted to design a neural network that could learn the XOR function, and while we were able to do so using non-linear functions and fixing the values ourselves, it was a far more challenging, impossible problem for the early neural networks that attempted to learn it. Part of the reason was that we needed the non-linear functions (such as tanh or reLu),  however, more importantly the early networks didn't have any form of back-propagation which was essential for the neural networks to be able to learn even what might the almost trivial XOR function. Without being able to pass the error backwards and impact all the layers of the net, it was impossible for the XOR function to be learned.

**2. Load / Manipulate the Boston Housing Dataset**

The first step will be to read in the Boston Housing data from the CSV

In [30]:
import numpy as np
import pandas as pd

boston_df = pd.read_csv("https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv")
boston_df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

To look at the dimensions of the dataset, we can just look at the shape of the dataframe.



In [24]:
print(boston_df.shape)


(506, 14)


Which shows that we have 14 different features we're interested in ("crim", "nox", "ptratio", etc) and 506 rows of data.


To split the available data into a learning, validation and test set, we can use scikit's train_test_split() function.

In [27]:
from sklearn.model_selection import train_test_split

# using an 80/20 split since that seemed recommended as a safe bet
# https://stackoverflow.com/questions/13610074/is-there-a-rule-of-thumb-for-how-to-divide-a-dataset-into-training-and-validatio
train, test = train_test_split(boston_df, test_size=0.2)
test, validation = train_test_split(test, test_size=0.2)

print(train.shape)
print(test.shape)
print(validation.shape)


(404, 14)
(81, 14)
(21, 14)


If we wanted to add a new feature to this data set, one option would be to

In [32]:
import tensorflow as tf


KeyError: ignored